frontend/src/components/elements/ArrowTable/ArrowTable.tsx

```diff
@@ -0,0 +1,177 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React, { Fragment, PureComponent, ReactNode, SFC } from "react"
+import { Table as ReactTable } from "reactstrap"
+import { Map as ImmutableMap } from "immutable"
+import { toFormattedString } from "lib/format"
+import withFullScreenWrapper from "hocs/withFullScreenWrapper"
+import TheArrowTable from "./TheArrowTable"
+
+export interface Props {
+  element: ImmutableMap<string, any>
+  width: number
+}
+
+export class ArrowTable extends PureComponent<Props> {
+  public render = (): ReactNode => {
+    const uuid: string = this.props.element.get("uuid")
+    const headerColumns: Uint8Array = this.props.element.get("headerColumns")
+    const headerRows: Uint8Array = this.props.element.get("headerRows")
+    const data: Uint8Array = this.props.element.get("data")
+    const table = new TheArrowTable(uuid, headerColumns, headerRows, data)
+
+    const styles: string = this.props.element.get("styles")
+    const hasNoData = table.headerRows === table.rows
+
+    const captionText: string = this.props.element.get("caption")
+    const caption = captionText ? <caption>{captionText}</caption> : null
+
+    return (
+      <div className="streamlit-table stTable">
+        <style>{styles}</style>
+        <ReactTable id={"T_" + uuid} className={hasNoData ? "empty-table" : ""}>
+          {caption}
+          <thead>
+            <TableRows
+              table={table}
+              header={true}
+              headerRows={table.headerRows}
+              rows={table.rows}
+              columns={table.columns}
+            />
+          </thead>
+          <tbody>
+            {hasNoData ? (
+              <tr>
+                <td colSpan={table.columns || 1}>empty</td>
+              </tr>
+            ) : (
+              <TableRows
+                table={table}
+                header={false}
+                headerRows={table.headerRows}
+                rows={table.rows}
+                columns={table.columns}
+              />
+            )}
+          </tbody>
+        </ReactTable>
+      </div>
+    )
+  }
+}
+
+/**
+ * Purely functional component returning a list of rows.
+ *
+ * table      - The table to display.
+ * header     - Whether to display the header.
+ * headerRows - Number of rows in the header.
+ * rows       - Number of rows in the table (header + data).
+ * columns    - Number of colums in the table.
+ */
+
+export interface TableRowsProps {
+  table: TheArrowTable
+  header: boolean
+  headerRows: number
+  rows: number
+  columns: number
+}
+
+const TableRows: SFC<TableRowsProps> = props => {
+  const { table, header, headerRows, rows, columns } = props
+  const startRow = header ? 0 : headerRows
+  const endRow = header ? headerRows : rows
+  const rowArray = []
+
+  for (let rowIndex = startRow; rowIndex < endRow; rowIndex++) {
```

Comment:
Nit... there is not functional alternative? Like using `map`?

Reply:
I don't like this part also. Any ideas?